Onboard Temperature Calibration [WIP]

bugobliterator opened the pull request:
Temperature Calibration Scheme for calibrating sensors. The module will be running onboard and processing data realtime.

#### TODOs:
Parameters for post and pre calibration stage not yet saved, add mechanism to store them.
Currently only gyro, include accel and baro.
Integrate with Events interface

#### Current Results:
Gyro Temperature Cal Temperature vs Gyro Bias plots (Temperature is on reference offset scale):
![gyro0_x](https://cloud.githubusercontent.com/assets/5697288/22174650/b99b6852-e009-11e6-92a5-2f25ddebaee9.jpg)
![gyro0_y](https://cloud.githubusercontent.com/assets/5697288/22174651/bdab91d8-e009-11e6-9d3c-1061ec26796e.jpg)
![gyro0_z](https://cloud.githubusercontent.com/assets/5697288/22174653/c111f448-e009-11e6-9ab9-ecfe34c9a63f.jpg)


`tempcal start` inside nsh will kick off temperature cal.

LorenzMeier commented on src/modules/tempcal/matrix_alg.cpp:
```diff
@@ -0,0 +1,394 @@
+ *    @returns                multiplied matrix i.e. A*B
+ */
+
+float *mat_mul(float *A, float *B, uint8_t n)
```
If we compare this to https://github.com/PX4/Matrix, what parts of this standalone Matrix implementation can be replaced by the standard? I would much prefer to extend Matrix with whatever is missing and not have any math classes in here.

jgoppert commented:
Is anyone online to give me an overview of the online temp sampling strategy?

bugobliterator commented:
I am online, which mode do you prefer for comm.

On Sat, 21 Jan 2017 at 9:10 PM James Goppert

<
mailto:James Goppert <notifications@github.com>
> wrote:

a, pre, code, a:link, body { word-wrap: break-word !important; }

Is anyone online to give me an overview of the online temp sampling strategy?

—

You are receiving this because you authored the thread.

Reply to this email directly,
https://github.com/PX4/Firmware/pull/6413#issuecomment-274268955
, or
https://github.com/notifications/unsubscribe-auth/AFbvCBqvlUqkg5ba_IjiGQfVnIONiPL_ks5rUibggaJpZM4LqEKe
.

jgoppert commented:
Google hangout?  james.goppert@gmail.com

jgoppert commented:
@bugobliterator and I came up with a short term plan:

1. Switch math over to matrix. He has it using matrix in the beginning but was unsure about the output and switched to known working code he had. If we can find any errors in the matrix inverse in Matrix through this process we will fix it.
2. Implement interval averaging for data collection to avoid over weighting one section of the fit.
3. Compare results to least sq. solution result.

Fit interface:
1. User enters start deg C, the fit starts when start temp bin is reached
2. Users enters stop deg C and the fit stops when stop bin is reached.
3. User enters number of bins for fit or it is hardcoded to 100 etc.. This will be important for comparison with least square solution using pseudoinverse onboard since we will need a matrix of this size which must be of static size.

Long term support for online adaptation:
Fixed bins are created and each bin is initialized to pre-calibrated data. While the user flies, they are updating the temperature bin they are in. Some consideration for the gyro bias estimation in the IEKF/ EKF2 is needed here, maybe already addressed in temp calibration document.

bugobliterator commented:
@jgoppert @LorenzMeier I have changed to using Matrix functions in polyfit and converted it into a template with changeable order of functions, hence it can be used for anything relating to polynomial/curve/spline fitting. So do you think its a good idea to move it to the matrix library as well for anybody to use?

LorenzMeier commented:
Ok. Please integrate with this PR next: https://github.com/PX4/Firmware/pull/6396 and enable the full operation. I would propose that:

  * user sends a temperature calibration command
  * calibrator stores an "enabled" param in memory
  * user cold-soaks the board
  * calibrator starts taking samples 2 seconds after boot if enabled param is set (we need eventually motion detection)
  * calibrator runs until reaching steady-state temperature (e.g. one minute within 0.5 deg or so)

This is a bit different from min and max as we likely need to work with what the user gives us (best effort). It would also be good to be able to set min and max as additional gates - if they are not reached, the calibration will be considered invalid.

We would set those to relatively small values though (e.g. -5 to 35) by default to users can reach them.

jgoppert commented:
@bugobliterator yes, I think we should add it to Matrix. Let's pull it in after we get this temp call stuff wrapped up.

LorenzMeier commented:
@bugobliterator Where is this standing?

bugobliterator commented:
Discussion and PR moved to https://github.com/PX4/Firmware/pull/6425
